crates/sui-core/src/authority/authority_store_pruner.rs

```diff
@@ -535,6 +552,29 @@ impl AuthorityStorePruner {
         Ok(Some(sst_file))
     }
 
+    fn pruning_tick_duration_ms(epoch_duration_ms: u64) -> u64 {
+        min(epoch_duration_ms / 2, 60 * 1000)
+    }
+
+    fn smoothed_max_eligible_checkpoint_number(
+        pruned_checkpoint: CheckpointSequenceNumber,
+        max_eligible_checkpoint: CheckpointSequenceNumber,
+        epoch_duration_ms: u64,
+    ) -> CheckpointSequenceNumber {
+        if max_eligible_checkpoint == 0 {
+            return max_eligible_checkpoint;
+        }
+        let num_intervals = epoch_duration_ms
+            .checked_div(Self::pruning_tick_duration_ms(epoch_duration_ms))
+            .unwrap_or(1);
+        let delta = max_eligible_checkpoint
```

Comment:
Have you tried, say in PTN, how does the smooth option behave? It appears that this delta will become smaller and smaller when `pruned_checkpoint` move towards `max_eligible_checkpoint`, so we may not able to prune everything at the end of the epoch.

Also, it seems that for authority with `num_epochs_to_retain == 0`, when smooth is on, it will accumulate `num_intervals` number of checkpoints and then all go for pruning, which also doesn't feel desirable.

I think ideally, we want to calculate the delta only once, upon discovering a new epoch, and use that throughout the epoch.

Reply:
thanks for the review!
* initial max_eligible_checkpoint is just the last executed executed checkpoint, so it consistently increasing and ,although intervals might not be fully uniform, they won't decrease at the end of the epoch. Filtering for the epoch itself is done here: https://github.com/MystenLabs/sui/blob/7b59c441fa05fe329c703f122ed57ebe84b52b84/crates/sui-core/src/authority/authority_store_pruner.rs#L407. Maybe I should rename the variable in this block to avoid confusion
* for num_epochs_to_retain == 0 smoother is explicitly disabled https://github.com/MystenLabs/sui/pull/16997/files#diff-293733010660563ec43cd7a59de0596c258b01287b60f24e09ff6aeaac213708R318